app/api.go

```diff
@@ -85,89 +92,112 @@ func (lm *LearningMaterialAPI) getOrCreateChallengeEnv() http.HandlerFunc {
 			ErrorResponse(w)
 			return
 		}
-		cc, err := client.GetChallenge(chals)
+		cr, err := client.GetClientRequest(chals)
 
-		//Create a new challenge
+		//Create a new Environment
 		if err != nil {
-			if client.RequestMade() >= lm.conf.API.MaxRequest {
-				ErrorResponse(w) //todo create maxrequest error page
+			if client.RequestMade() >= lm.conf.API.ClientMaxRequest {
+				ClientTooManyRequests(w) //todo create maxrequest error page
 				return
 			}
-			go lm.CreateChallengeEnv(client, mux.Vars(r)["chals"])
+			go lm.CreateEnvironment(client, chals)
+			WaitingResponse(w)
 			return
 		}
 
 		//Check for error while creating the environment
 		select {
-		case err = <-cc.err:
+		case err = <-cr.err:
 			ErrorResponse(w)
 			return
 		default:
 		}
 
-		if !cc.isReady {
-			log.Info().Msgf("[NOT READY] Environment [%s] for the client [%s]", chals, client.ID())
+		if !cr.isReady {
+			//log.Info().Msgf("[NOT READY] Environment [%s] for the client [%s]", chals, client.ID())
 			WaitingResponse(w)
 			return
 		}
 
-		log.Info().Msgf("[READY] Environment [%s] for the client [%s]", chals, client.ID())
+		log.Info().Msgf("[READY] Client Request [%s] for the client [%s]", chals, client.ID())
 
-		authC := http.Cookie{Name: "GUAC_AUTH", Value: cc.guacCookie, Path: "/guacamole/"}
+		authC := http.Cookie{Name: "GUAC_AUTH", Value: cr.guacCookie, Path: "/guacamole/"}
 		http.SetCookie(w, &authC)
-		host := fmt.Sprintf("http://%s:%d/guacamole", lm.conf.Host, cc.guacPort)
+		host := fmt.Sprintf("/guacamole/?%s=%s", requestedChallenges, chals)
+		time.Sleep(5 * time.Second)
```

Comment:
`time.Sleep` seems very weak here and also it blocks main thread. We may think to change it with something else, let's talk it at meeting.